src/Workspaces/CoreTest/SemanticModelReuse/SemanticModelReuseTests.cs

```diff
@@ -0,0 +1,368 @@
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.UnitTests.SemanticModelReuse
+{
+    [UseExportProvider]
+    public class SemanticModelReuseTests
+    {
+        private static Document CreateDocument(string code, string language)
+        {
+            var solution = new AdhocWorkspace().CurrentSolution;
+            var projectId = ProjectId.CreateNewId();
+            var project = solution.AddProject(projectId, "Project", "Project.dll", language).GetProject(projectId);
+
+            return project.AddMetadataReference(TestReferences.NetFx.v4_0_30319.mscorlib)
+                          .AddDocument("Document", SourceText.From(code));
+        }
+
+        #region C# tests
+
+        [Fact]
+        public async Task NullBodyReturnsNormalSemanticModel1_CSharp()
+        {
+            var document = CreateDocument("", LanguageNames.CSharp);
+
+            // trying to get a model for null should return a non-speculative model
+            var model = await document.ReuseExistingSpeculativeModelAsync(null, CancellationToken.None);
+            Assert.False(model.IsSpeculativeSemanticModel);
+        }
+
+        [Fact]
+        public async Task NullBodyReturnsNormalSemanticModel2_CSharp()
+        {
+            var source = "class C { void M() { return; } }";
+            var document = CreateDocument(source, LanguageNames.CSharp);
+
+            // Even if we've primed things with a real location, getting a semantic model for null should return a
+            // non-speculative model.
+            var model1 = await document.ReuseExistingSpeculativeModelAsync(source.IndexOf("return"), CancellationToken.None);
+            var model2 = await document.ReuseExistingSpeculativeModelAsync(null, CancellationToken.None);
+            Assert.False(model1.IsSpeculativeSemanticModel);
+            Assert.False(model2.IsSpeculativeSemanticModel);
+        }
+
+        [Fact]
+        public async Task SameSyntaxTreeReturnsNonSpeculativeModel_CSharp()
+        {
+            var source = "class C { void M() { return; } }";
+            var document = CreateDocument(source, LanguageNames.CSharp);
+
+            // First call will prime the cache to point at the real semantic model.  The next call will also use the
+            // same syntax tree, so it should get the same semantic model.
+            var model1 = await document.ReuseExistingSpeculativeModelAsync(source.IndexOf("return"), CancellationToken.None);
+            var model2 = await document.ReuseExistingSpeculativeModelAsync(source.IndexOf("return"), CancellationToken.None);
+            Assert.False(model1.IsSpeculativeSemanticModel);
+            Assert.False(model2.IsSpeculativeSemanticModel);
+
+            // Should be the same models.
+            Assert.Equal(model1, model2);
```

Comment:
There's Assert.Same() if you want to explicitly state object instances versus any (potential) Equals override.